deduplicator/src/deduplicator.rs

```diff
@@ -0,0 +1,313 @@
+use std::mem::drop;
+use std::path::PathBuf;
+use std::thread;
+
+use crossbeam_channel as channel;
+use importer_tools::Address;
+use itertools::Itertools;
+use prog_rs::prelude::*;
+use rusqlite::{Connection, DropBehavior};
+
+use crate::db_hashes::{DbHashes, HashIterItem};
+use crate::dedupe::{hash_address, is_duplicate};
+use crate::utils::is_constraint_violation_error;
+
+const CHANNEL_SIZES: usize = 1000;
+
+pub struct Deduplicator {
+    db: DbHashes,
+}
+
+impl Deduplicator {
+    pub fn new(output_path: PathBuf) -> rusqlite::Result<Self> {
+        Ok(Self {
+            db: DbHashes::new(output_path)?,
+        })
+    }
+
+    pub fn get_db_inserter<F, R>(&mut self, filter: F, ranking: R) -> rusqlite::Result<DbInserter>
+    where
+        F: Fn(&Address) -> bool + Clone + Send + 'static,
+        R: Fn(&Address) -> f64 + Clone + Send + 'static,
+    {
+        Ok(DbInserter::new(self.db.get_conn()?, filter, ranking))
+    }
+
+    pub fn compute_duplicates(&mut self) -> rusqlite::Result<()> {
+        println!("Build index on hashes");
+        self.db.create_hashes_index()?;
+
+        // --- Query collisions from DB
+        let count_addresses_before = self.db.count_addresses()?;
+        let count_hashes = self.db.count_hashes()?;
+
+        println!(
+            "Compute hash collisions ({} addresses, {} hashes)",
+            count_addresses_before, count_hashes
+        );
+
+        let conn_get_collisions = self.db.get_conn()?;
+        let mut sorted_hashes = DbHashes::get_sorted_hashes(&conn_get_collisions)?;
+
+        // Eliminate false positives in parallel using following pipeline:
+        //
+        // [     col_sender     ] main thread
+        //            |
+        //            |  (address, rank)
+        //            v
+        // [    col_receiver    ]
+        // [         |||        ] worker threads
+        // [     del_sender     ]
+        //            |
+        //            |  (address, rank, hashes)
+        //            v
+        // [    del_receiver    ] writter thread
+
+        let nb_workers: usize = num_cpus::get() - 2;
+        let (col_sender, col_receiver) = channel::bounded::<Vec<HashIterItem>>(CHANNEL_SIZES);
+        let (del_sender, del_receiver) = channel::bounded(CHANNEL_SIZES);
+
+        // --- Init worker threads
+
+        for _ in 0..nb_workers {
+            let col_receiver = col_receiver.clone();
+            let del_sender = del_sender.clone();
+
+            thread::spawn(move || {
+                for mut pack in col_receiver {
+                    if pack.len() > 1000 {
+                        eprintln!("Skipping pack of length {}", pack.len());
```

Comment:
Why?

Reply:
If we get collisions of size > 1000 we may want to implement a new way to handle that to avoid critical performance issues. Currently we iterate over all feasible pairs to check if they are true duplicates, if the number of addresses of the collisions is 1000 there is about 1 million of such pairs.

Reply:
I detailed the intention behind a comment:

```rust
// In practice this should not happen often, however in the case where this
// issue is raised, it would be necessary to implement a specific way of
// handling big packs (for example by computing more accurate hashes in
// RAM).
```